autotools doc PR: No depends_on('m4') with depends_on('autoconf')

bernhardkaindl opened the pull request:
* depends_on('m4') is obsolete with depends_on('autoconf'):
  - Remove `m4` from the example depends_on() lines for the autoreconf phase.
  - Change the branch used as example from develop to master as it is
    far more common in the packages of spack's builtin repo.

* Fix the wrong info that libtoolize and aclocal are run explicitly
  in the autoreconf phase by default. autoreconf calls these internally
  as needed, thus autotools.py also does not call them directly.

* Add that autoreconf() also adds -I<aclocal-prefix>/share/aclocal.

* Add an example how to set autoreconf_extra_args.

* Add an example of a custom autoreconf phase for running autogen.sh.

haampie commented:
See https://github.com/spack/spack/pull/26079#pullrequestreview-759570995

bernhardkaindl commented:
> See https://github.com/spack/spack/pull/26079#pullrequestreview-759570995

Thanks, updated as per suggestion of @haampie in https://github.com/spack/spack/pull/26079#pullrequestreview-759570995 :

> You should use the public api and limit to direct build type dependencies: spec.dependencies(deptype='build') and then never use in as it seems to lead to false positives indeed:
```py
from:
-        missing = [x for x in autotools if x not in spec]
to:
+        build_deps = [d.name for d in spec.dependencies(deptype='build')]
+        missing = [x for x in needed_dependencies if x not in build_deps]
```
Another included improvement: 

As `autoconf` has `m4` as run-dependency, m4 does not have to be added as `depends_on()` explicitly.
`depends_on(autoconf)` implies that m4 (indirectly) is provided as well, so packages do not have to add it explicitly to each package:
```diff
-        autotools = ['m4', 'autoconf', 'automake', 'libtool']
+        autotools = ['autoconf', 'automake', 'libtool']
```
This removes the need of having this in all packages building `@master`:
```py
+    depends_on('m4',       when='@master', type='build')
```
Just this is enough (tested)
```py
+    depends_on('autoconf', when='@master', type='build')
+    depends_on('automake', when='@master', type='build')
+    depends_on('libtool',  when='@master', type='build')
```
I'll see how the CI tests go. Yes, indeed, I think we'll likely want/need a test for this check. Any hints and help would be welcome.

haampie commented:
Could you update the docs here too: https://spack.readthedocs.io/en/latest/build_systems/autotoolspackage.html?highlight=autotools#build-system-dependencies they promote adding m4 as a direct build dependency.

bernhardkaindl committed:
autotoolspackage.rst: No depends_on('m4') with depends_on('autoconf')

- depends_on('m4') is obsolete when depends_on('autoconf') is present:
  - Remove `m4` from the example depends_on() lines for the autoreconf phase.
  - Change the branch used as example from develop to master as it is
    far more common in the packages of spack's builtin repo.

- Fix the wrong info that libtoolize and aclocal are run explicitly
  in the autoreconf phase by default. autoreconf calls these internally
  as needed, thus autotools.py also does not call them directly.

- Add that autoreconf() also adds -I<aclocal-prefix>/share/aclocal.

- Add an example how to set autoreconf_extra_args.

- Add an example of a custom autoreconf phase for running autogen.sh.

bernhardkaindl commented:
> Could you update the docs here too: https://spack.readthedocs.io/en/latest/build_systems/autotoolspackage.html?highlight=autotools#build-system-dependencies they promote adding m4 as a direct build dependency.

done, I've also worked on other findings for the autoreconf phase. Updated of the commit message:
```py
   Documentation update:
    
    Update the documentation accordingly and fix the wrong info that
    libtoolize and aclocal are run explicitly before calling autoreconf:
    autoreconf calls these internally as needed, thus autotools.py also
    does not call them directly. Fix this misleading info in the doc.
    
    Add that autoreconf() also adds -I<aclocal-prefix>/share/aclocal.
    
    Also add an example how to set autoreconf_extra_args when needed.

    Remove `m4` from the example depends_on() lines for the autoreconf
    tools, but describe that it is needed to have it for spack version
    0.16.x and older for now.
    
    Change the branch used as example from develop to master as it is
    far more common in the packages of spack's builtin repo.
    
    Added an example of a custom autoreconf phase for running autogen.sh.
```

haampie commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -112,20 +112,47 @@ phase runs:
+   depends_on('automake', type='build', when='@master')
+   depends_on('libtool',  type='build', when='@master')
+
+When the package has ``autoconf`` as dependenency, and ``m4`` is only used
+as the script interpreter of ``autoconf`` and not for other purposes, it
+does not have to be listed explicitly and can be removedd as it is not needed.
+Spack versions 0.16.x and lower may wrongly terminate their autoreconf phase
+if it is not listed as an explicit dependency.
```
```suggestion
```
I think we can skip this and target the docs to (new) users of the current version of Spack. In fact `m4` is not mentioned anywhere in this page (well, except in the `autoreconf_extra_args` example). If people add `m4` anyways we can catch it during review of their PR.

bernhardkaindl replied:
@haampie Done, rebased to develop and now it is only a doc PR. Removed the other labels. Ready for review.

haampie commented:
LGTM except that comment on the docs

haampie commented:
Can you fix the conflict? Then it should be just a docs PR, right?

bernhardkaindl commented:
> Can you fix the conflict? Then it should be just a docs PR, right?

Yes, will do. Then it just a docs PR, correct.

haampie commented:
Cool, I'll rebase my PR on develop/this then!

haampie commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -112,20 +112,45 @@ phase runs:
+   depends_on('automake', type='build', when='@master')
+   depends_on('libtool',  type='build', when='@master')
+
+When the package has ``autoconf`` as dependenency, and ``m4`` is only used
+as the script interpreter of ``autoconf`` and not for other purposes, it
+does not have to be listed explicitly and can be removedd as it is not needed.
```
This sentence is still here though, I'm not very attached to my opinion about removing it. But in this sentence it's somwhat ambiguous what "it" refers to, m4 or autoconf. How about:

```suggestion
It is typically redundant to list the ``m4`` macro processor package as a
dependency, since ``autoconf`` already depends on it.
```

bernhardkaindl replied:
@haampie  Applied your suggestion, thanks!

> It is typically redundant to list the ``m4`` macro processor package as a
dependency, since ``autoconf`` already depends on it.

bernhardkaindl committed:
Applyied suggestion to shorten the m4/autoconf paragraph

Thanks, much better!

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>
